nupic/research/frameworks/vernon/core_mixins/distributed.py

```diff
@@ -0,0 +1,137 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import copy
+
+import torch.distributed as dist
+from torch import multiprocessing
+
+__all__ = [
+    "Distributed",
+]
+
+
+class Distributed:
+    """
+    Extends the BaseExperiment to work in distributed scenarios.
+
+    Any class inheriting from Distributed should:
+    - Implement aggregate_results and aggregate_pre_experiment_results
+      if results from different processes need to be aggregated.
+    - Wrap models with DistributedDataParallel
+    - Use DistributedSamplers when using dataloaders
+    - Update samplers with `sampler.set_epoch()`
+    """
+    def setup_experiment(self, config):
+        """
+        :param config: Dictionary containing the configuration parameters
+
+            - distributed: Whether or not to use Pytorch Distributed training
+            - backend: Pytorch Distributed backend ("nccl", "gloo")
+                    Default: nccl
+            - world_size: Total number of processes participating
+            - rank: Rank of the current process
+        """
+        distributed = config.get("distributed", False)
+        rank = config.get("rank", 0)
+
+        if distributed and rank != 0:
+            # Only enable logs from first process
+            config = copy.copy(config)
+            config["disable_logger"] = True
+
+        super().setup_experiment(config)
+
+        self.distributed = distributed
+        self.rank = rank
+
+        # CUDA runtime does not support the fork start method.
+        # See https://pytorch.org/docs/stable/notes/multiprocessing.html
+        multiprocessing.set_start_method("spawn", force=True)
+
+        if self.distributed:
+            dist_url = config.get("dist_url", "tcp://127.0.0.1:54321")
```

Comment:
Just a side comment, not a change request: there is a small but so far neglectable chance this port will be taken. Might become an issue if extending framework to more users to throw an error if that is the case or choose a free port instead.